executor.go

```diff
@@ -0,0 +1,253 @@
+		mkdirMutexMap:        map[string]*sync.Mutex{},
+		executionHashes:      map[string]context.Context{},
+		executionHashesMutex: sync.Mutex{},
+	}
+	e.Options(opts...)
+	return e
+}
+
+func (e *Executor) Options(opts ...ExecutorOption) {
+	for _, opt := range opts {
+		opt(e)
+	}
+}
+
+func WithDir(dir string) ExecutorOption {
+	return func(e *Executor) {
+		e.Dir = dir
+	}
+}
+
+func WithEntrypoint(entrypoint string) ExecutorOption {
+	return func(e *Executor) {
+		e.Entrypoint = entrypoint
+	}
+}
+
+func WithTempDir(tempDir TempDir) ExecutorOption {
+	return func(e *Executor) {
+		e.TempDir = tempDir
+	}
+}
+
+func WithForce(force bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Force = force
+	}
+}
+
+func WithForceAll(forceAll bool) ExecutorOption {
+	return func(e *Executor) {
+		e.ForceAll = forceAll
+	}
+}
+
+func WithInsecure(insecure bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Insecure = insecure
+	}
+}
+
+func WithDownload(download bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Download = download
+	}
+}
+
+func WithOffline(offline bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Offline = offline
+	}
+}
+
+func WithTimeout(timeout time.Duration) ExecutorOption {
+	return func(e *Executor) {
+		e.Timeout = timeout
+	}
+}
+
+func WithWatch(watch bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Watch = watch
+	}
+}
+
+func WithVerbose(verbose bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Verbose = verbose
+	}
+}
+
+func WithSilent(silent bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Silent = silent
+	}
+}
+
+func WithAssumeYes(assumeYes bool) ExecutorOption {
+	return func(e *Executor) {
+		e.AssumeYes = assumeYes
+	}
+}
+
+func WithDry(dry bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Dry = dry
+	}
+}
+
+func WithSummary(summary bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Summary = summary
+	}
+}
+
+func WithParallel(parallel bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Parallel = parallel
+	}
+}
+
+func WithColor(color bool) ExecutorOption {
+	return func(e *Executor) {
+		e.Color = color
+	}
+}
+
+func WithConcurrency(concurrency int) ExecutorOption {
+	return func(e *Executor) {
+		e.Concurrency = concurrency
+	}
+}
+
+func WithInterval(interval time.Duration) ExecutorOption {
+	return func(e *Executor) {
+		e.Interval = interval
+	}
+}
+
+func WithOutputStyle(outputStyle ast.Output) ExecutorOption {
+	return func(e *Executor) {
+		e.OutputStyle = outputStyle
+	}
+}
+
+func WithTaskSorter(sorter sort.Sorter) ExecutorOption {
+	return func(e *Executor) {
+		e.TaskSorter = sorter
+	}
+}
+
+func WithStdin(stdin io.Reader) ExecutorOption {
+	return func(e *Executor) {
+		e.Stdin = stdin
+	}
+}
+
+func WithStdout(stdout io.Writer) ExecutorOption {
+	return func(e *Executor) {
+		e.Stdout = stdout
+	}
+}
+
+func WithStderr(stderr io.Writer) ExecutorOption {
+	return func(e *Executor) {
+		e.Stderr = stderr
+	}
+}
```

Comment:
In theory we could have an additional function that accepts all three, for shortness. Optional suggestion, though.

```go
func WithIO(stdin io.Reader, stdout, stderr io.Writer) ExecutorOption {
	return func(e *Executor) {
		e.Stdin = stdin
		e.Stdout = stdout
		e.Stderr = stderr
	}
}
```

Reply:
To me, it makes more sense to keep these separate if you're passing different values into each of them. Sometimes you might only want to pass in 1 or 2 of them and leave the others blank. However, I think if you're setting them all to the same value, we could have something like this:

```go
func WithIO(rw io.ReadWriter) ExecutorOption {
	return func(e *Executor) {
		e.Stdin = rw
		e.Stdout = rw
		e.Stderr = rw
	}
}
```

This would be super handy in the new tests I've been writing where we want to write everything to a single buffer.